### ISSUE TITLE
JSONPath params are always converted to String

### ISSUE THREAD

**Issue by juanpablo-santos** [jenkinsci/generic-webhook-trigger-plugin#7]
Hi,

Currently, if a JSONPath expression selects a node rather than a leaf, then no matter what, it gets converted to a String. This is because on `VariableResolver#addPostContentParameters`, the call to the `filter` methods uses a `toString` on the resolved variable (which, in JSONPath can also be a map, if the expression resolves to a node instead of a leaf)

For large pipelines, which use a lot of JSONPath params, it would be nice to have one or two variables representing JSONPath nodes, instead of having one variable for each needed value (as a side effect, this would also allow to access the whole webhook payload). F.ex, inside the pipeline is really easy to do something like
```
"""
Dear email template,
<a href="$sender.email">$sender.full_name"</a> has come to known blablabla
"""
```
and it's much more manageable if the job has only one or a couple of JSONPath params rather than 10 or 12.

It seems that to achieve this only requires a couple of changes on `VariableResolver`:

1) on [VariableResolver#L62](https://github.com/jenkinsci/generic-webhook-trigger-plugin/blob/master/src/main/java/org/jenkinsci/plugins/gwt/VariablesResolver.java#L62), the call to `filter` should be changed to:
```
map.put(gv.getKey() + "_" + i, filter(o, gv.getRegexpFilter()));
```

2) replace the [`filter` method](https://github.com/jenkinsci/generic-webhook-trigger-plugin/blob/master/src/main/java/org/jenkinsci/plugins/gwt/VariablesResolver.java#L118) with:
```
private Object filter(Object string, String regexpFilter) {
    if( string instanceof String ) {
        if (string == null || regexpFilter == null || regexpFilter.isEmpty()) {
          return string;
        }
        return string.replaceAll(regexpFilter, "");
    } else {
        LOGGER.info( "NOT filtering " + string )
        return object;
    }
}
```

As noted above, the ability to filter values on the parameter would be lost for nodes, although that functionality wouldn't be really useful for them anyway.

WDYT?

thanks,
juan pablo

**Comment by juanpablo-santos** [comment]
Hi,

having a closer look at this, it is not as straightforward as I thought initially. Will tinker a bit to see if it's possible to come with something not too obscure.

br,
juan pablo

**Comment by tomasbjerre** [comment]
Perhaps if a node is found from a JSONPath or XPath then all leafs in the node can be contributed in variables like:
```
$nodename_leafname1
$nodename_leafname2
$nodename_leafname3
...
```

**Comment by juanpablo-santos** [comment]
Hi Tomas,

I think I'll be able to have a PR for by tomorrow; I've already made the appropiate changes with their corresponding unit tests, but would like to test it on a live installation first to ensure evertyhing is fine. 

br,
juan pablo

**Comment by tomasbjerre** [comment]
I released this in 1.11 now!

**Comment by juanpablo-santos** [comment]
Cool, thanks! :-)
